### PR TITLE
add CloudCoverDetectionDataModule

### PR DESCRIPTION
**Pull request by robmarkcole**
Add datamodule for [CloudCoverDetection](https://github.com/torchgeo/torchgeo/blob/9289a5550818c662638085f4653b41d739f856f7/torchgeo/datasets/cloud_cover.py#L23C7-L23C26)

Do we want to normalise? e.g. 
```
    mean = torch.tensor(0)
    std = torch.tensor(10000)
```

It trains as-is however:
image

### CODE REVIEW

File: tests/datamodules/test_cloud_cover.py
Comment (adamjstewart) [inline tests/datamodules/test_cloud_cover.py:1]: My goal is to remove everything in `tests/datamodules`. Instead, we should add a `tests/conf/cloud_cover.yaml` and test this via SemanticSegmentationTask.

File: torchgeo/datamodules/cloud_cover.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        Args:
+            batch_size: Size of each mini-batch.
+            num_workers: Number of workers for parallel data loading.
+            val_split_pct: Percentage of the training data to reserve for validation.
+            **kwargs: Additional keyword arguments passed to
+                :class:`~torchgeo.datasets.CloudCoverDetection`.
+        """
+        if not 0 < val_split_pct < 1:
+            msg = 'val_split_pct must be between 0 and 1.'
+            raise ValueError(msg)
```
Comment (adamjstewart) [inline torchgeo/datamodules/cloud_cover.py]: This will be tricky to get coverage of, can probably just remove

**Comment by adamjstewart** [comment]
We definitely want to normalize to the 0-1 range. The default is to divide by 255, but if the images have larger values, we can divide by a larger number.

File: tests/conf/cloud_cover.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    task: 'binary'
+    loss: 'bce'
+    num_filters: 1
+data:
+  class_path: CloudCoverDetectionDataModule
+  init_args:
+    batch_size: 1
+    val_split_pct: 0.2
+  dict_kwargs:
+    root: 'tests/data/cloud_cover_detection'
```
Comment (adamjstewart) [inline tests/conf/cloud_cover.yaml]: This is the same as `tests/data/ref_cloud_cover_detection_challenge_v1`? We can rename this directory if it's confusing.

**Comment by robmarkcole** [comment]
@adamjstewart there is a bit of a mix of naming amongst `cloud_cover.py`, `CloudCoverDetectionDataModule` and cloud-cover-detection-challenge in  https://staging.source.coop/radiantearth/cloud-cover-detection-challenge I feel the latter is the most unique and could be the best choice to adopt?

File: tests/conf/cloud_cover.yaml (hunk 2 above)
Comment (robmarkcole) [reply tests/conf/cloud_cover.yaml]: rename to `cloud_cover_detection_challenge` ?

**Comment by adamjstewart** [comment]
Perhaps @klwetstone has opinions on the name of the dataset since she created https://source.coop/radiantearth/cloud-cover-detection-challenge.

I would like to differentiate between the name of the competition and the name of the dataset. "Sentinel-2 Cloud Cover Segmentation" is probably correct, although a bit long.

File: tests/conf/cloud_cover.yaml (hunk 2 above)
Comment (adamjstewart) [reply tests/conf/cloud_cover.yaml]: Let's finalize the name of the dataset and dataset file first. For now, you can use `tests/data/ref_cloud_cover_detection_challenge_v1` to get the tests to pass without any file duplication.

**Comment by robmarkcole** [comment]
Sentinel-2 Cloud Cover Segmentation could apply to other datasets, e.g. Cloudsen12. I assume `ref_cloud_cover_detection_challenge_v1` is a truncation of `Radiant Earth Foundation Sentinel-2 Cloud Cover Segmentation Dataset (Version 1)` although the use of `ref` is a bit confusing. `radian-earth-cloud-detection-v1` would make sense to me

**Comment by robmarkcole** [comment]
the dataset expects `train/test` folders but the current test approach has them as `public/private` - @adamjstewart which approach do you want?

**Comment by adamjstewart** [comment]
What does the real data look like? Did they change the directory hierarchy again? The test data should always match the real data.

**Comment by robmarkcole** [comment]
The real data uses `test/train` - not sure if this has changed?
image

The testing data
image

**Comment by adamjstewart** [comment]
Must have changed. Can you fix this in a separate PR so we can backport the fix to 0.8.1? This PR introduces a new datamodule, so it has to wait until 0.9.0.

**Comment by robmarkcole** [comment]
@adamjstewart we can [see that](https://github.com/torchgeo/torchgeo/blob/109d83d4ba7cad4256fce74deea047641e6b88d1/torchgeo/datasets/cloud_cover.py#L168) although the [remote dir](https://staging.source.coop/radiantearth/cloud-cover-detection-challenge/final) is called `public` we sync to a `train` dir - I assume you want to keep this behaviour? This does mean if we do 
```
data:
  class_path: CloudCoverDetectionDataModule
  init_args:
    batch_size: 1
    val_split_pct: 0.2
  dict_kwargs:
    root: 'tests/data/ref_cloud_cover_detection_challenge_v1'
``` 
the paths are incorrect, since the data has not been syncd and renamed

**Comment by adamjstewart** [comment]
Oh, I think I understand now. I'm fine with either. The `split` parameter should take in 'train' or 'test', but I don't care what the download directory is called. It might be simpler if we keep the public/private names in case someone manually downloads and wants to use it in TorchGeo.

**Comment by robmarkcole** [comment]
@adamjstewart agreed - you want a seperate PR for that or do here?

**Comment by adamjstewart** [comment]
Hmm, maybe separate.

**Comment by robmarkcole** [comment]
Pending https://github.com/torchgeo/torchgeo/pull/3145

File: docs/api/datamodules.rst [hunk 3]
```diff
@@ -82,6 +82,11 @@ ChaBuD
 
 .. autoclass:: ChaBuDDataModule
 
+Cloud Cover Detection
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (adamjstewart) [inline docs/api/datamodules.rst]:
```suggestion
Cloud Cover Detection
^^^^^^^^^^^^^^^^^^^^^
```

File: tests/conf/cloud_cover.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    backbone: 'resnet18'
+    in_channels: 4
+    task: 'binary'
+    loss: 'bce'
+    num_filters: 1
+data:
+  class_path: CloudCoverDetectionDataModule
+  init_args:
+    batch_size: 1
+    val_split_pct: 0.2
```
Comment (adamjstewart) [inline tests/conf/cloud_cover.yaml]: Could also use 0.5 to guarantee that val is not empty, we don't have very many fake images compared to the real dataset.

File: torchgeo/datamodules/cloud_cover.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+        Args:
+            batch_size: Size of each mini-batch.
+            num_workers: Number of workers for parallel data loading.
+            val_split_pct: Percentage of the training data to reserve for validation.
+            **kwargs: Additional keyword arguments passed to
+                :class:`~torchgeo.datasets.CloudCoverDetection`.
+        """
+        self.mean = torch.tensor(0.0)
+        self.std = torch.tensor(10000.0)
```
Comment (adamjstewart) [inline torchgeo/datamodules/cloud_cover.py]: Can we put these at the class level like the base class?

**Comment by robmarkcole** [comment]
@adamjstewart the masks contain raw uint16 values `([2732, …])` - might be why I struggled to train a model. Addressed now

**Comment by adamjstewart** [comment]
Hmm, I just checked `public/train_labels/*.tif` and all files are in the range [0, 1]. Which files did you encounter 2732 in?

**Comment by robmarkcole** [comment]
@adamjstewart thanks for checking, the issue was in the mock dataset

**Comment by adamjstewart** [comment]
Still curious about train accuracy. Want to run a quick training run to see if accuracy has improved with normalization?

**Comment by robmarkcole** [comment]
Batch size 8
```
task = SemanticSegmentationTask(
    model="unet",
    backbone="resnet18",
    weights=True,
    in_channels=4,
    num_classes=2,
    # loss="focal",
    # labels=['background', 'cloud'],
    class_weights=Tensor([1, 10]),
    lr=1e-4,
)
```
Possibly overfitting after around 10 epochs
image
